codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+ * Channel#read()}, which follows the default behavior of channels in Netty. Applications are
+ * free to close the channel in response to such events if they don't have use for any queued
+ * messages.
+ *
+ * <p>{@link ChannelConfig#setMaxMessagesPerRead(int)} and {@link
+ * ChannelConfig#setAutoRead(boolean)} are supported.
+ */
+public class Http2MultiplexCodec extends ChannelDuplexHandler {
+    private static final Http2FrameLogger HTTP2_FRAME_LOGGER
+        = new Http2FrameLogger(INFO, Http2MultiplexCodec.class);
+
+    private final ChannelHandler streamHandler;
+    private final EventLoopGroup streamGroup;
+    private final Http2ConnectionHandler http2Handler;
+    private final Http2Connection.PropertyKey streamInfoKey;
+    private final List<StreamInfo> streamsToFireChildReadComplete = new ArrayList<StreamInfo>();
+    private ChannelHandlerContext ctx;
+    private ChannelHandlerContext http2HandlerCtx;
+
+    /**
+     * Construct a new handler whose child channels run in the same event loop as this handler.
+     *
+     * @param server {@code true} this is a server
+     * @param streamHandler the handler added to channels for remotely-created streams. It must be
+     *     {@link ChannelHandler.Sharable}.
+     */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        this(server, streamHandler, null);
+    }
+
+    /**
+     * Construct a new handler whose child channels run in a different event loop.
+     *
+     * @param server {@code true} this is a server
+     * @param streamHandler the handler added to channels for remotely-created streams. It must be
+     *     {@link ChannelHandler.Sharable}.
+     * @param streamGroup event loop for registering child channels
+     */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup) {
+        this(server, streamHandler, streamGroup, new DefaultHttp2FrameWriter());
+    }
+
+    Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup, Http2FrameWriter frameWriter) {
+        if (!streamHandler.getClass().isAnnotationPresent(Sharable.class)) {
+            throw new IllegalArgumentException("streamHandler must be Sharable");
+        }
+        this.streamHandler = streamHandler;
+        this.streamGroup = streamGroup;
+        Http2Connection connection = new DefaultHttp2Connection(server);
+        frameWriter = new Http2OutboundFrameLogger(frameWriter, HTTP2_FRAME_LOGGER);
+        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
+        Http2FrameReader reader = new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), HTTP2_FRAME_LOGGER);
+        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
+        decoder.frameListener(new FrameListener());
```

Comment:
> > Even if I fixed it now, I would have zero confidence that it wouldn't leak this in the future.

> I know this code is new, and interfaces are very "fresh", but I think we should just focus on the current state of the code.

My concern wasn't related to the code being new. The concern is that any change can break the property. And I have limited faith that a review would notice such things 50% of the time.

>> Given that this class is the sole owner of the objects involved, I could argue that in some ways it isn't even really leaking this.

> Agreed having internal classes limits exposure for bad things happening...but coordination of internals of the class is still necessary. If escaping can be avoided that is just 1 less thing to worry about.

I wasn't meaning that. I was talking about more "what is the definition of 'this escaping'".

These examples escape:
```java
class A {
  A(Escaper e) {
    e.set(this);
  }
}
```

```java
class B {
  B() {
    nonFinalMethod();
  }
  protected void nonFinalMethod() {}
}
```

Both of those can lead to problems without trying too hard.

But this one I'd argue is debatable in severity:
```java
class C {
  List<Object> l = new ArrayList<Object>();
  C() {
    l.add(this):
  }
}
```

And that's because C is the root all the objects that contain its pointer, and those objects aren't escaped. This is the case it appears Http2MultipleCodec is in. And [AbstractChannel](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractChannel.java#L83). And while it is nice to reduce such graphs of objects, especially during construction, I don't see too much gain in avoiding it with a Builder or factory method at this point, especially since I can choose to simply not run those lines of code in the constructor and delay to `handlerAdded()` if it becomes a problem.